File: sort.py
#!/usr/bin/env python3

from random import randint
from typing import List

def main() -> None:
  # Setup
  data: List[int] = [randint(0, 99) for _ in range(100)]

  # Process
  qsort(data)

  # Output
  print('\t'.join(str(x) for x in data))


def qsort(data: List[int]) -> None:
  if len(data) > 1:
    pivot: int = data[0]
    left: List[int] = [x for x in data[1:] if x < pivot]
    right: List[int] = [x for x in data[1:] if x >= pivot]
    qsort(left)
    qsort(right)
    i: int = 0
    for value in left:
      data[i] = value
      i += 1
    data[i] = pivot
    i += 1
    for value in right:
      data[i] = value
      i += 1


if __name__ == '__main__':
  main()

File: sort.rb
#!/usr/bin/env ruby

def qsort(data)
  return if data.length < 2
  pivot = data.first
  left = data[1..].select { |x| x < pivot }
  right = data[1..].select { |x| x >= pivot }
  qsort(left)
  qsort(right)
  i = 0
  left.map { |x| data[i] = x; i += 1 }
  data[i] = pivot
  right.map { |x| data[i] = x; i += 1 }
end

# Setup
data = (1..100).map {|_| rand(100)}

# Process
qsort(data)

#Output
puts data.join("\t")
